### PR TITLE
app: switch to typescript

### PR DESCRIPTION
**Pull request by buck54321**
Change all files to typescript and enforce typing during builds. Shared types are defined in registry.ts.

### CODE REVIEW

File: client/webserver/site/src/js/markets.ts [hunk 1]
```diff
@@ -1977,16 +2109,16 @@ export default class MarketsPage extends BasePage {
...
       else bttn.classList.remove('selected')
     }
-    const { candleCaches, cfg } = this.market
+    const { candleCaches, cfg, baseUnitInfo, quoteUnitInfo } = this.market
     const cache = candleCaches[this.candleDur]
     if (cache) {
       this.depthChart.hide()
       this.candleChart.show()
-      this.candleChart.setCandles(cache, cfg)
+      this.candleChart.setCandles(cache, cfg, baseUnitInfo, quoteUnitInfo)
```
Comment (buck54321) [inline client/webserver/site/src/js/markets.ts:2121]: This seems to have been a bug.

File: client/webserver/site/src/js/markets.ts [hunk 2]
```diff
@@ -743,7 +874,7 @@ export default class MarketsPage extends BasePage {
     this.maxOrderUpdateCounter++
     const mkt = this.market
     const baseWallet = app().assets[mkt.base.id].wallet
-    if (baseWallet.available < mkt.cfg.lotsize) {
+    if (baseWallet.balance.available < mkt.cfg.lotsize) {
```
Comment (buck54321) [inline client/webserver/site/src/js/markets.ts:877]: Seems to have also been a bug.

File: client/webserver/site/src/js/registry.ts [hunk 3]
```diff
@@ -0,0 +1,506 @@
...
+    localeDiscrepancies: () => void
+  }
+}
+
+export interface Exchange {
+  host: string
+  acctID: string
+  markets: Record<string, Market>
+  assets: Record<number, Asset>
+  connected: boolean
```
Comment (chappjc) [inline client/webserver/site/src/js/registry.ts:16]: Oh. My. Gawd.  It's beautiful.  ❤️ ❤️

File: client/webserver/site/tsconfig.json [hunk 4]
```diff
@@ -0,0 +1,11 @@
+{
+  "compilerOptions": {
+    "outDir": "./dist/",
+    "noImplicitAny": true,
```
[the remark this answers is not in the document]
Comment (chappjc) [reply client/webserver/site/tsconfig.json:4]: Looks like we're already satisfying `--noUnusedLocals --noUnusedParameters`.  Maybe include those too?

File: client/webserver/site/package-lock.json [hunk 5]
```diff
@@ -62,9 +67,9 @@
       }
     },
     "node_modules/@babel/compat-data": {
-      "version": "7.16.8",
-      "resolved": "https://registry.npmjs.org/@babel/compat-data/-/compat-data-7.16.8.tgz",
-      "integrity": "sha512-m7OkX0IdKLKPpBlJtF561YJal5y/jyI5fNfWbPxh2D/nbzzGI4qRyrD8xO2jB24u7l+5I2a43scCG2IrfjC50Q==",
+      "version": "7.17.7",
```
Comment (chappjc) [inline client/webserver/site/package-lock.json:70]: It makes little difference, but could `npm dedup` the file too.

**Review by chappjc — COMMENTED** [review]
This is very exciting work!

This is not a proper review, just some comments in scanning through the overall changes.

File: client/webserver/site/webpack/prod.js [hunk 6]
```diff
@@ -20,16 +16,18 @@ module.exports = merge(common, {
   module: {
     rules: [
       {
-        test: /\.js$/,
+        test: /\.ts$/,
         exclude: /node_modules/,
         use: {
+          // babel-loader does not fail on type errors. ts-loader does, but we
+          // probably still want to transpile (right?).
```
[the remark this answers is not in the document]
Comment (buck54321) [reply client/webserver/site/webpack/prod.js:23]: I actually have it just running `tsc` before `webpack`.

```
"build": "./node_modules/.bin/tsc && webpack --config webpack/prod.js --progress --color",
```

File: client/webserver/site/src/js/registry.ts (hunk 3 above)
Comment (buck54321) [reply client/webserver/site/src/js/registry.ts:16]: I had a feeling you'd like TypeScript.

**Review by martonp — COMMENTED** [review]
This is great! Will be much nicer to work on the UI from now on. 

What do you think about turning `strictNullChecks` on? There will be many errors to fix.. I can give it a shot in another PR if you don't want to. I think it can be a pain sometimes, but it will make things a lot safer.

**Review by JoeGruffins — APPROVED** [review]

File: client/webserver/site/webpack/dev.js [hunk 7]
```diff
@@ -1,11 +1,14 @@
...
-const ESLintPlugin = require('eslint-webpack-plugin')
 
 module.exports = merge(common, {
   mode: 'development',
-  plugins: [new ESLintPlugin({
-    formatter: 'stylish'
-  })],
+  module: {
+    rules: [{
+      test: /\.tsx?$/,
```
[the remark this answers is not in the document]
Comment (buck54321) [reply client/webserver/site/webpack/dev.js]: removed `x?`

File: client/webserver/site/package-lock.json (hunk 5 above)
Comment (buck54321) [reply client/webserver/site/package-lock.json:70]: done

**Review by martonp — APPROVED** [review]

**Review by chappjc — APPROVED** [review]

**Pull request merged**